zebra-chain/src/transparent/utxo.rs

```diff
@@ -22,32 +22,94 @@ pub struct Utxo {
     pub from_coinbase: bool,
 }
 
-/// Compute an index of newly created transparent outputs, given a block and a
+/// A [`Utxo`], and the index of its transaction within its block.
+///
+/// This extra index is used to check that spends come after outputs,
+/// when a new output and its spend are both in the same block.
+///
+/// The extra index is only used during block verification,
+/// so it does not need to be sent to the state.
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(
+    any(test, feature = "proptest-impl"),
+    derive(proptest_derive::Arbitrary)
+)]
+pub struct OrderedUtxo {
+    /// An unspent transaction output.
+    pub utxo: Utxo,
+    /// The index of the transaction that created the output, in the block at `height`.
+    ///
+    /// Used to make sure that transaction can only spend outputs
+    /// that were created earlier in the chain.
+    ///
+    /// Note: this is different from `OutPoint.index`,
+    /// which is the index of the output in its transaction.
+    pub tx_index_in_block: usize,
+}
+
+impl OrderedUtxo {
+    /// Create a new ordered UTXO from its fields.
+    pub fn new(
+        output: transparent::Output,
+        height: block::Height,
+        from_coinbase: bool,
+        tx_index_in_block: usize,
+    ) -> OrderedUtxo {
+        let utxo = Utxo {
+            output,
+            height,
+            from_coinbase,
+        };
+
+        OrderedUtxo {
+            utxo,
+            tx_index_in_block,
+        }
+    }
+}
+
+/// Compute an index of [`Utxo`]s, given an index of [`OrderedUtxo`]s.
+pub fn utxos_from_ordered_utxos(
+    ordered_utxos: &HashMap<transparent::OutPoint, OrderedUtxo>,
+) -> HashMap<transparent::OutPoint, Utxo> {
+    ordered_utxos
+        .iter()
+        .map(|(out_point, ordered_utxo)| (*out_point, ordered_utxo.utxo.clone()))
+        .collect()
+}
```

Comment:
Minor: Would it make sense to make this take ownership of the `ordered_utxos`? It would avoid the `clone` :thinking: 

```suggestion
pub fn utxos_from_ordered_utxos(
    ordered_utxos: HashMap<transparent::OutPoint, OrderedUtxo>,
) -> HashMap<transparent::OutPoint, Utxo> {
    ordered_utxos
        .into_iter()
        .map(|(out_point, ordered_utxo)| (out_point, ordered_utxo.utxo))
        .collect()
}
```

Reply:
Fixed in 24f66df7